token_provider.go

```diff
@@ -66,14 +75,27 @@ func NewAutoRefreshTokenProvider(tokenClient *TokenClient) *AutoRefreshTokenProv
 }
 
 func (t *AutoRefreshTokenProvider) GetToken() (Token, error) {
+	var unlockOnce sync.Once
+
+	t.tokenMutex.RLock()
+	defer unlockOnce.Do(t.tokenMutex.RUnlock)
+
 	if t.token == nil || t.token.IsExpired() {
+		unlockOnce.Do(t.tokenMutex.RUnlock)
```

Comment:
Extract this to a function or reorganize the code so you don't have to unlock the token in multiple places.

```suggestion
	t.tokenMutex.RLock()
	token := t.token
	t.tokenMutex.RUnlock()

	if token == nil || token.IsExpired() {
```

Reply:
Also, the earliest return in this case is inverted. You should return token earlier than t.refresh(), as the former is constant.

```go
if token != nil && !token.IsExpired() {
	return token, nil
}

return t.refresh()
```